File: FSUIKit/Classes/FSButtonView.m
////
////  FSButtonView.m
////  FBRetainCycleDetector
////
////  Created by FudonFuchina on 2020/2/23.
////
//
//#import "FSButtonView.h"
//
//@implementation FSButtonView {
//    UIView      *_backView;
//}
//
//- (instancetype)initWithFrame:(CGRect)frame {
//    self = [super initWithFrame:frame];
//    if (self) {
//        [self buttonDesignViews];
//    }
//    return self;
//}
//
//- (void)buttonDesignViews {
//    _backView = [[UIView alloc] initWithFrame:self.bounds];
//    [self addSubview:_backView];
//    UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tapClick)];
//    [_backView addGestureRecognizer:tap];
//}
//
//- (void)tapClick {
//    if (self.tap) {
//        self.tap(self);
//    }
//}
//
//- (void)layoutSubviews {
//    [super layoutSubviews];
//    _button.frame = CGRectMake(5, self.frame.size.height / 2 - 18, self.frame.size.width - 10, 36);
//}
//
//- (UIButton *)button {
//    if (!_button) {
//        _button = [UIButton buttonWithType:UIButtonTypeSystem];
//        _button.frame = CGRectMake(5, self.frame.size.height / 2 - 18, self.frame.size.width - 10, 36);
//        _button.layer.cornerRadius = 18;
//        [_button addTarget:self action:@selector(tapClick) forControlEvents:UIControlEventTouchUpInside];
//        [self addSubview:_button];
//    }
//    return _button;
//}
//
///*
//// Only override drawRect: if you perform custom drawing.
//// An empty implementation adversely affects performance during animation.
//- (void)drawRect:(CGRect)rect {
//    // Drawing code
//}
//*/
//
//@end

File: FSUIKit/Classes/Swift/FSButtonView.swift
// FSButtonViewS.swift
// Translated from FSButtonView.h/m

import UIKit

public class FSButtonView: UIView {
    
    private var _backView: UIView!
    private var _button: UIButton?
    
    public var tap: ((FSButtonView) -> Void)?
    
    public var button: UIButton {
        if _button == nil {
            let btn = UIButton(type: .system)
            btn.frame = CGRect(x: 5, y: frame.height / 2 - 18, width: frame.width - 10, height: 36)
            btn.layer.cornerRadius = 18
            btn.addTarget(self, action: #selector(tapClick), for: .touchUpInside)
            addSubview(btn)
            _button = btn
        }
        return _button!
    }
    
    public override init(frame: CGRect) {
        super.init(frame: frame)
        buttonDesignViews()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        buttonDesignViews()
    }
    
    private func buttonDesignViews() {
        _backView = UIView(frame: bounds)
        addSubview(_backView)
        
        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(tapClick))
        _backView.addGestureRecognizer(tapGesture)
    }
    
    @objc private func tapClick() {
        tap?(self)
    }
    
    public override func layoutSubviews() {
        super.layoutSubviews()
        _button?.frame = CGRect(x: 5, y: frame.height / 2 - 18, width: frame.width - 10, height: 36)
    }
}
